src/nsrt_learning/strips_learning/gen_to_spec_learner.py

```diff
@@ -229,119 +277,187 @@ def _backchain_one_pass(
+        best_pnad = None
+        best_ground_pnad = None
+        for pnad in pnads:
+            param_opt, opt_vars = pnad.option_spec
+            assert param_opt == segment.get_option().parent
+            isub = dict(zip(opt_vars, option_objs))
+            # Handle keep effect substitution correctly
+            # NOTE: code copied directly from _recompute_datastores.
+            if segment in pnad.seg_to_keep_effects_sub:
+                # If there are any variables only in the keep effects,
+                # their mappings should be put into isub, since their
+                # grounding is underconstrained by the segment itself.
+                keep_eff_sub = pnad.seg_to_keep_effects_sub[segment]
+                for var in pnad.op.parameters:
+                    if var in keep_eff_sub:
+                        assert var not in isub
+                        isub[var] = keep_eff_sub[var]
+
+            # Loop over all groundings.
+            for ground_pnad in utils.all_ground_operators_given_partial(
+                    pnad.op, objects, isub):
+                if not ground_pnad.preconditions.issubset(segment.init_atoms):
                     continue
-            else:
-                if not ground_op.add_effects.issubset(segment.final_atoms):
-                    continue
-                if not necessary_add_effects.issubset(ground_op.add_effects):
-                    continue
-            return ground_op
-        return None
+                if ground_eff_subset_necessary_eff:
+                    if not ground_pnad.add_effects.issubset(
+                            necessary_add_effects):
+                        continue
+                else:
+                    if not ground_pnad.add_effects.issubset(
+                            segment.final_atoms):
+                        continue
+                    if not necessary_add_effects.issubset(
+                            ground_pnad.add_effects):
+                        continue
+                # This ground PNAD covers this segment. Score it!
+                score = self._score_segment_ground_op_match(
+                    segment, ground_pnad)
+                if score < best_score:  # we want a closer match
+                    best_score = score
+                    best_pnad = pnad
+                    best_ground_pnad = ground_pnad
+        return (best_pnad, best_ground_pnad)
 
     def _try_specializing_pnad(
         self,
         necessary_add_effects: Set[GroundAtom],
-        pnad: PartialNSRTAndDatastore,
+        pnads: List[PartialNSRTAndDatastore],
         segment: Segment,
-    ) -> Optional[PartialNSRTAndDatastore]:
-        """Given a PNAD and some necessary add effects that the PNAD must
-        achieve, try to make the PNAD's add effects more specific
+    ) -> Optional[Tuple[PartialNSRTAndDatastore, PartialNSRTAndDatastore]]:
+        """Given a list of PNAD and some necessary add effects that one of the
```

Comment:
"list of PNAD" -> "list of PNADs" (but this might be obsolete if you accept my suggestion above that this method should take in a single PNAD)